text/0162-refactoring-support.md

```diff
@@ -0,0 +1,622 @@
+this is sometimes impractical or conflicts with good software engineering
+practices. For instance, you may want to consolidate duplicated code across
+different CDK applications into a single reusable construct (often called an L3
+construct). Introducing a new node for the L3 construct in the construct tree
+will rename the logical IDs of the resources in that subtree.
+
+Also, you might need to move resources within the tree for better readability or
+between stacks to isolate concerns. Accidental renames have also caused issues
+for customers in the past. Perhaps even worse, if you depend on a third-party
+construct library, you are not in control of the logical IDs of those resources.
+If the library changes the logical IDs from one version to another, you will be
+affected without any action on your part.
+
+To address all these problems, the CDK CLI now automatically detects these
+cases, and refactors the stack on your behalf, using the new CloudFormation
+stack refactoring API. This brings more flexibility for developers, and reduces
+the risk of accidental changes that lead to resource renaming.
+
+### How it works
+
+When you run `cdk deploy`, the CLI will compare the templates in the cloud
+assembly with the templates in the deployed stack. If it detects that a resource
+has been moved or renamed, it will automatically perform the refactoring, and
+then proceed with the deployment.
+
+For example, suppose your CDK application has a single stack, called `MyStack`,
+containing an S3 bucket, a CloudFront distribution and a Lambda function. The
+construct tree looks like this (L1 constructs omitted for brevity):
+
+    App
+    └ MyStack
+      ├ Bucket
+      ├ Distribution
+      └ Function
+
+Now suppose you make the following changes, after having deployed it to your AWS
+account:
+
+- Rename the bucket from `Bucket` to the more descriptive name `Origin`.
+- Create a new L3 construct called `Website` that groups the bucket and the
+  distribution, to make this pattern reusable in different applications.
+- Move the web-related constructs (now under the `Website` L3) construct to a
+  new stack called `Web`, for better separation of concerns.
+- Rename the original stack to `Service`, to better reflect its new specific
+  role in the application.
+
+The refactored construct tree looks like this:
+
+    App
+    ├ Web
+    │  └ Website
+    │    ├ Origin
+    │    └ Distribution
+    └ Service
+      └ Function
+
+Even though none of the resources have changed, their paths have
+(from `MyStack/Bucket/Resource` to `Web/Website/Origin/Resource` etc.) Since the
+CDK computes the logical IDs of the resources from their path in the tree, all
+three resources will have their logical IDs changed.
+
+If you run `cdk deploy` now, by default the CLI will detect these changes and
+present you with a selection prompt:
+
+    The following resources were moved or renamed:
```

Comment:
This describes behavior after implementation. It would be helpful to also include an example of behavior before the implementation (what the CDK currently does), to better illustrate to users the problem we are solving.